src/v/cluster/cloud_storage_size_reducer.h

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "cluster/cluster_utils.h"
+#include "cluster/controller_service.h"
+#include "cluster/logger.h"
+#include "cluster/members_table.h"
+#include "cluster/topic_table.h"
+#include "cluster/topic_table_partition_generator.h"
+#include "config/node_config.h"
+#include "rpc/connection_cache.h"
+
+namespace cluster {
+
+class cloud_storage_size_reducer_exception : public std::runtime_error {
+public:
+    explicit cloud_storage_size_reducer_exception(const std::string& m)
+      : std::runtime_error(m) {}
+};
+
+/*
+ * This utility class performs a map-reduce operation accross the cluster
+ * in order to determine the total number of cloud storage bytes used by
+ * all partitions.
+ *
+ * It iterates across the current topic table in batche. For each batch,
+ * the following sequence of operations is performed:
+ * 1. Find the first live replica of each partition in the batch
+ * 2. Prepare `cloud_storage_usage` RPC requests for each node in the cluster.
+ * The request will contain the partitions being queried by shard.
+ * 3. Reduce the reponses and update the accumulator.
+ *
+ * One important thing to note is that this operation requires topic table
+ * stability (i.e. the topic table does not change in the meantime). This
+ * property is enfored by the `topic_table_partition_generator`, which will
+ * throw if the topic table has been updated during the iteration. If such
+ * a change does happen, it's treated as a retryable error.
+ */
+class cloud_storage_size_reducer {
+public:
+    static constexpr uint8_t default_retries_allowed = 3;
+
+    cloud_storage_size_reducer(
+      ss::sharded<topic_table>& topic_table,
+      ss::sharded<members_table>& members_table,
+      ss::sharded<rpc::connection_cache>& conn_cache,
+      size_t batch_size = topic_table_partition_generator::default_batch_size,
+      uint8_t retries_allowed = 3)
+      : _topic_table(std::ref(topic_table))
+      , _members_table(std::ref(members_table))
+      , _conn_cache{conn_cache}
+      , _batch_size{batch_size}
+      , _retries_allowed{retries_allowed} {}
+
+    ss::future<std::optional<uint64_t>> reduce() {
+        try {
+            co_return co_await do_reduce();
+        } catch (const topic_table_partition_generator_exception& e) {
+            vlog(
+              clusterlog.warn,
```

Comment:
these types of warnings would make CI fail when they are retry-able and in the end the reduction may succeed. Probably better to log them as debug until the number of retries has been exhausted

Reply:
Fixed